crates/ruff_python_parser/src/typing.rs

```diff
@@ -93,7 +93,17 @@ fn parse_simple_type_annotation(
 
 fn parse_complex_type_annotation(string_expr: &ExprStringLiteral) -> AnnotationParseResult {
     let mut parsed = parse_expression(string_expr.value.to_str())?;
-    relocate_expr(parsed.expr_mut(), string_expr.range());
+    // Remove quotes from nested forward reference type annotations
+    let range_excluding_quotes =
+        if let Expr::StringLiteral(ExprStringLiteral { ref value, .. }) = *parsed.syntax.body {
+            let string_parts = value.as_slice();
+            let start = string_parts[0].flags.opener_len();
+            let end = string_parts[string_parts.len() - 1].flags.closer_len();
+            string_expr.range().add_start(start).sub_end(end)
+        } else {
+            string_expr.range()
+        };
+    relocate_expr(parsed.expr_mut(), range_excluding_quotes);
```

Comment:
The way it would work with simple type annotations is that it first parses the outer string and then the inner string on the next call:
```
parse_simple_type_annotation: "'int'" (10..15)
parse_simple_type_annotation: "int" (11..14)
```

In the case of complex annotations, it seems like the range is always going to be the one without the inner quotes:
```
parse_complex_type_annotation: "'int'" (10..18)
parse_complex_type_annotation: "int" (10..18)
```